db/version_set.cc

```diff
@@ -345,6 +347,385 @@ class FilePicker {
+    if (!search_ended_) {
+      // REVISIT
+      // Prefetch Level 0 table data to avoid cache miss if possible.
+      // As of now, only PlainTableReader and CuckooTableReader do any
+      // prefetching. This may not be necessary anymore once we implement
+      // batching in those table readers
+      for (unsigned int i = 0; i < (*level_files_brief_)[0].num_files; ++i) {
+        auto* r = (*level_files_brief_)[0].files[i].fd.table_reader;
+        if (r) {
+          for (auto iter = range_->begin(); iter != range_->end(); ++iter) {
+            r->Prepare(iter->ikey);
+          }
+        }
+      }
+    }
+  }
+
+  int GetCurrentLevel() const { return curr_level_; }
+
+  FdWithKeyRange* GetNextFile() {
+    while (!search_ended_) {
+      // Start searching next level.
+      if (mget_iter_ == current_level_range_.end()) {
+        search_ended_ = !PrepareNextLevel();
+        continue;
+      } else {
+        if (maybe_repeat_key_) {
+          maybe_repeat_key_ = false;
+          // Check if we found the final value for the last key in the
+          // previous lookup range. If we did, then there's no need to look
+          // any further for that key, so advance mget_iter_. Else, keep
+          // mget_iter_ positioned on that key so we look it up again in
+          // the next file
+          if (current_level_range_.CheckKeyDone(mget_iter_)) {
+            ++mget_iter_;
+          }
+        }
+        // prev_mget_iter_end_ will become the start key for the next file
+        // lookup
+        prev_mget_iter_end_ = mget_iter_;
+      }
+
+      FdWithKeyRange* f = nullptr;
+      FilePickerContextWrapper file_picker_iter(mget_iter_, file_picker_ctx_);
+      size_t curr_file_index = (mget_iter_ != current_level_range_.end())
+                                   ? file_picker_iter->curr_index_in_curr_level
+                                   : curr_file_level_->num_files;
+      bool file_hit = false;
+      int cmp_largest = -1;
+      MultiGetRange next_file_range(current_level_range_, prev_mget_iter_end_,
+                                    current_level_range_.end());
+      if (curr_file_index >= curr_file_level_->num_files) {
+        search_ended_ = !PrepareNextLevel();
+        continue;
+      }
+      hit_file_level_ = curr_level_;
+      is_hit_file_last_in_level_ = file_picker_iter->curr_index_in_curr_level ==
+                                   curr_file_level_->num_files - 1;
+      // Loops over all files in current level.
+      while (mget_iter_ != current_level_range_.end() &&
+             (file_picker_iter->curr_index_in_curr_level == curr_file_index ||
+              !file_hit)) {
+        struct FilePickerContext& fp_ctx = *file_picker_iter;
+        f = &curr_file_level_->files[fp_ctx.curr_index_in_curr_level];
+        Slice& user_key = mget_iter_->ukey;
+
+        // Do key range filtering of files or/and fractional cascading if:
+        // (1) not all the files are in level 0, or
```

Comment:
Do we support fractional cascading fro L0 at all?

Reply:
No. IIUC, its not possible because we have to go through every L0 flie.